pyccel/ast/c_concepts.py

```diff
@@ -7,14 +7,30 @@
 Module representing object address.
 """
 
-from .basic import TypedAstNode, PyccelAstNode
+from pyccel.utilities.metaclasses import Singleton
+from .basic     import TypedAstNode, PyccelAstNode
+from .datatypes import DataType
 from .literals  import LiteralString
 
 __all__ = ('CMacro',
+           'CStackArray',
            'CStringExpression',
            'ObjectAddress',
            'PointerCast')
 
+#------------------------------------------------------------------------------
+
+class CStackArray(DataType, metaclass=Singleton):
+    """
+    A data type representing an array allocated on the stack.
+
+    A data type representing an array allocated on the stack.
+    E.g. `float a[4];`
+    """
+    __slots__ = ()
+    _name = 'c_stackarray'
+
+#------------------------------------------------------------------------------
```

Comment:
Written like this, I'm not sure. I added it because I needed to create a `void API_pointers[n_classes];` and this seemed to be the cleanest way of doing it.

Conceptually though it is already possible to allocate on the stack. See : https://github.com/pyccel/pyccel/blob/devel/docs/decorators.md#stack-array

(See also https://fortran-lang.discourse.group/t/why-stack-is-faster-than-heap-and-what-exactly-is-stack/2130/6 for more details about why you might want to use stack instead of heap). I think I allocate tuples on the stack but I'm not sure (I think there was some disagreement about the right way to handle this). In my opinion, tuples should be on the stack as they are usually small objects (I think the counter argument was that this can't be guaranteed). Lists tend to be bigger so they should be allocated on the heap

Reply:
I guess both arguments are valid, in a perfect world, there would be a "check first and do" for the available stack memory and the size of the tuple coming from the input,  or at least "try and recover", but i doubt both approaches would be efficient.

Reply:
I agree, it is also tricky to implement. I'm not aware of a (ideally multi-platform) command that tells you the available stack memory. And there is no "try and recover" mechanism for this. Heap is therefore the safest and `@stack_array` can always be used to allocate on the stack if the user thinks there is an advantage